Bellfruit Black Box driver

SomeRandomGuyIdk opened the pull request:
- Driver for Bellfruit Black Box fruit machines, 15 games playable with barebones layouts
- Added a device for EM reels, which mpu1.cpp also now uses
- Support for looping samples, couple new ones and cleanup in fruitsamples.cpp

SomeRandomGuyIdk commented:
https://github.com/SomeRandomGuyIdk/mame/blob/ebd7a0e20a385b78b29e915f62207fa502955564/src/mame/bfm/bfm_blackbox.cpp#L1064-L1065
Is there any way to get access to the internal RAM of a 6802? I'm pretty sure it needs initializing to a nonzero value here, so I had to use the regular 6800 and add RAM to the driver's map instead.

SomeRandomGuyIdk commented:
https://github.com/SomeRandomGuyIdk/mame/blob/ebd7a0e20a385b78b29e915f62207fa502955564/src/mame/bfm/bfm_blackbox.h#L24-L27
The I/O for outputs is a bit funny, it works by putting 4-bit data onto PIA port B and then accessing an address mapped to a device to write the data. Most games write to the addresses, but of course bb_gspin happens to read them instead, so I came up with these macros to make code that does the exact same thing for both reads and writes a bit cleaner. Could I be missing some simpler way to make a write just use the read function in cases where the written data is not used?

SomeRandomGuyIdk commented:
https://github.com/SomeRandomGuyIdk/mame/blob/ebd7a0e20a385b78b29e915f62207fa502955564/src/mame/bfm/bfm_blackbox.cpp#L404-L414
There's one game here that has a fixed beeper as part of it's sound, the beep in this case is a modulated sine wave, so not trivial to synthesize and I don't know the circuit either. So I made a sample (>1kB in size) and stuck it in fruitsamples.cpp, is that OK? There shouldn't be many fruit techs left using analog circuits for sound after this...

james-wallace-ghub commented:
Would [beep.cpp](https://github.com/mamedev/mame/blob/master/src/devices/sound/beep.cpp) in the sound devices work in this circumstance? It's meant to be a PC beeper type thing, but somewhat less sophisticated than DACs etc. It might be a step up from playing a sample in this case.
> https://github.com/SomeRandomGuyIdk/mame/blob/ebd7a0e20a385b78b29e915f62207fa502955564/src/mame/bfm/bfm_blackbox.cpp#L404-L414 There's one game here that has a fixed beeper as part of it's sound, the beep in this case is a modulated sine wave, so not trivial to synthesize and I don't know the circuit either. So I made a sample (>1kB in size) and stuck it in fruitsamples.cpp, is that OK? There shouldn't be many fruit techs left using analog circuits for sound after this...

SomeRandomGuyIdk commented:
beep.cpp does square waves, doesn't help with sine waves unfortunately. Anyway, I decided to stop being lazy and made the code generate the beep sample now, hopefully that should work.

cuavas commented on src/mame/bfm/bfm_blackbox.h:
```diff
@@ -0,0 +1,299 @@
+#include "sound/beep.h"
+#include "sound/spkrdev.h"
+
+#define OUT_READ(name) uint8_t name##_r(offs_t offset) { name##_w(offset, 0); return 0; }
+#define OUT_READ_T(name) template <unsigned T> uint8_t name##_r(offs_t offset) { name##_w<T>(offset, 0); return 0; }
+#define OUTPUT(name) rw(FUNC(name##_r), FUNC(name##_w))
+#define OUTPUT_T(name, T) rw(FUNC(name##_r<T>), FUNC(name##_w<T>))
```
Please don’t do this.  It just obfuscates things and makes debugging more difficult.  We’re trying to move aware from macros except in cases where they’re really needed.  We already demacrofied all the address maps.  This is regressive.

SomeRandomGuyIdk replied:
Fixed.

cuavas commented on src/mame/bfm/bfm_blackbox.h:
```diff
@@ -0,0 +1,299 @@
+
+**********************************************************************/
+
+#ifndef MAME_INCLUDES_BFM_BLACKBOX_H
+#define MAME_INCLUDES_BFM_BLACKBOX_H
```
This doesn’t match the file path.  Also, is there a good reason for having this stuff in a header?  Is it going to be inherited from another source file?  If not, the declarations should be in the .cpp file, and everything besides the driver definitions should be in an anonymous namespace.

SomeRandomGuyIdk replied:
Fixed. I just did it to split things up since the driver has a lot of classes, and will probably end up getting a few more once I add the stepper games.

cuavas commented on src/mame/bfm/bfm_blackbox.h:
```diff
@@ -0,0 +1,299 @@
+class blackbox_base_state : public driver_device
+{
+public:
+	blackbox_base_state(const machine_config &mconfig, device_type type, const char *tag) :
+		driver_device(mconfig, type, tag),
+		m_maincpu(*this, "maincpu"),
+		m_ram(*this, "ram"),
+		m_chute_timer(*this, "chute_timer"),
+		m_pia(*this, "pia"),
+		m_in_1800(*this, "IN1800"),
+		m_in_1800_en(*this, "IN1800_%u", 1U),
+		m_in_2000(*this, "IN2000"),
+		m_in_perc(*this, "PERCENTAGE"),
+		m_lamps(*this, "lamp%u", 0U),
+		m_digits(*this, "digit%u", 0U),
+		m_test_led(*this, "test%u", 1U),
+		m_samples(*this, "samples"),
+		m_nvram(*this, "nvram", 0x40, ENDIANNESS_BIG)
+	{ }
+
+	void blackbox_base(machine_config &config);
```
If this is an abstract class and abstract machine configuration, this stuff should be `protected`.

SomeRandomGuyIdk replied:
Fixed.

cuavas commented on src/mame/bfm/bfm_blackbox.h:
```diff
@@ -0,0 +1,299 @@
+class blackbox_em_base_state : public blackbox_base_state
+{
+public:
+	blackbox_em_base_state(const machine_config &mconfig, device_type type, const char *tag) :
+			blackbox_base_state(mconfig, type, tag),
+			m_reels(*this, "emreel%u", 1U)
+	{ }
+
+	void blackbox_em_base_20symbol(machine_config &config);
+	void blackbox_em_base_24symbol(machine_config &config);
```
Same for this.

SomeRandomGuyIdk replied:
Fixed.

cuavas reviewed (commented):
A 6802 depending on the state of RAM on start doesn’t sound right.  It’s static RAM, so you can’t depend on it containing any particular value.

I’m really not keen on having devices with their own sound samples.  I also don’t like the forced 65,536 steps of the existing reels, and I don’t think modelling new devices on that is a great idea.

happppp commented:
> https://github.com/SomeRandomGuyIdk/mame/blob/ebd7a0e20a385b78b29e915f62207fa502955564/src/mame/bfm/bfm_blackbox.cpp#L1064-L1065 Is there any way to get access to the internal RAM of a 6802? I'm pretty sure it needs initializing to a nonzero value here, so I had to use the regular 6800 and add RAM to the driver's map instead.

machine_start:

	for (int i = 0; i < 0x80; i++)
		m_maincpu->space(AS_PROGRAM).write_byte(i, 0xff);

(I'm putting 0xff there, makes more sense than 0x01)

Paul-Arnold commented on src/mame/bfm/bfm_blackbox.cpp:
```diff
@@ -0,0 +1,1331 @@
+	map(0x0500, 0x057f).OUTPUT_T(blackbox_em_admc_state::out_lamps, 34);
+	map(0x0580, 0x05ff).OUTPUT_T(blackbox_em_admc_state::out_lamps, 26);
+	map(0x0600, 0x067f).OUTPUT_T(blackbox_em_admc_state::out_lamps, 18);
+	map(0x0800, 0x0800).nopw(); // Writes 3, then 2 here on startup
```
That address is a 6850 uart.

SomeRandomGuyIdk replied:
Yep that checks out, thanks!

Paul-Arnold commented:
Black Box was actually Bell Fruits second cpu based technology.
Before that was “flat chassis”. IIRC it was 6800, 128 byte ram chip and I think two 2708 EPROMs.  Housed in a large slim aluminium case which it got its name from.
I’m not aware of any surviving machines or rom dumps :(

SomeRandomGuyIdk commented on src/mame/barcrest/mpu1.cpp:
```diff
@@ -412,6 +324,7 @@ void mpu1_state::mpu1(machine_config &config)
 	m_maincpu->set_addrmap(AS_PROGRAM, &mpu1_state::mpu1_map);
 
 	TIMER(config, "nmi").configure_periodic(FUNC(mpu1_state::nmi), attotime::from_hz(100)); // From AC zero crossing detector
+	subdevice<timer_device>("nmi")->set_start_delay(attotime::from_msec(1)); // Don't go to NMI at reset time
```
Fixed.

SomeRandomGuyIdk commented on src/mame/bfm/bfm_blackbox.cpp:
```diff
@@ -0,0 +1,1331 @@
+	// Check if switch is set high on 2-position games
+	if(perc_sw == 1 && !m_input_en[3] && !m_input_en[5]) return true;
+
+	return false;
```
Fixed.

SomeRandomGuyIdk commented on src/mame/bfm/bfm_blackbox.cpp:
```diff
@@ -0,0 +1,1331 @@
+	TIMER(config, "nmi").configure_periodic(FUNC(blackbox_base_state::nmi), attotime::from_hz(100)); // From AC zero crossing detector
+	TIMER(config, "irq").configure_periodic(FUNC(blackbox_base_state::irq), attotime::from_msec(80)); // 555 timer circuit
+	subdevice<timer_device>("nmi")->set_start_delay(attotime::from_msec(1)); // Don't take interrupts at reset time
+	subdevice<timer_device>("irq")->set_start_delay(attotime::from_msec(1));
```
Fixed.

SomeRandomGuyIdk commented on src/mame/bfm/bfm_blackbox.cpp:
```diff
@@ -0,0 +1,1331 @@
+	EM_REEL(config, m_reels[1], 24, attotime::from_usec(3000));
+	EM_REEL(config, m_reels[2], 24, attotime::from_usec(3000));
+	EM_REEL(config, m_reels[3], 24, attotime::from_usec(3000));
+}
```
Fixed.

SomeRandomGuyIdk commented:
> A 6802 depending on the state of RAM on start doesn’t sound right. It’s static RAM, so you can’t depend on it containing any particular value.

It's starting out with 0 that causes the problem, anything else works. There's a couple values the problematic function uses which it doesn't initialize in any way.

> Black Box was actually Bell Fruits second cpu based technology. Before that was “flat chassis”. IIRC it was 6800, 128 byte ram chip and I think two 2708 EPROMs. Housed in a large slim aluminium case which it got its name from. I’m not aware of any surviving machines or rom dumps :(

Interesting... MPU Mecca seems to have zero mentions of this thing. Could be possible that it's rare because it was primarily used for export games. I remember finding some Summit Coin 4040 tech there, the only photo of it was inside IIRC an Austrian machine.

Paul-Arnold commented:
> Interesting... MPU Mecca seems to have zero mentions of this thing. Could be possible that it's rare because it was primarily used for export games. I remember finding some Summit Coin 4040 tech there, the only photo of it was inside IIRC an Austrian machine.


I've not managed to find any info on it anywhere, not even a picture.
It was definitely used in the UK. I, and several others I know, remember repairing them. There's other BFM tech too which doesn't appear anywhere, most likely all lost now :(

I assume you've got the rom dumps and any other info that's in the downloads section of MPU Mecca/fruitemu ?

SomeRandomGuyIdk commented:
> I assume you've got the rom dumps and any other info that's in the downloads section of MPU Mecca/fruitemu ?

Yep, I paid the fiver to get access. Not easy to find things sometimes since it's just a big list of files, but there's quite a bit of useful stuff there.

james-wallace-ghub commented:
Not wanting to disrupt this discussion too much, but just to say we have a thread that we use to bring all this together outside of the pull request setup, simply because it can get a bit unwieldy for those doing reviews, and it may get lost when these are closed and merged.

https://github.com/orgs/mamedev/discussions/2

You'll see that there's different discussions on there about different aspects of FME in MAME, and I think if you haven't already, it'd be great to get your input there.

DavidHaywood commented:
@cuavas anything preventing this one from being merged?

cuavas commented:
Reviewing this now, sorry for the delay.

cuavas reviewed (changes requested):
Besides the stuff noted in bfm_blackbox.cpp itself, I’m still uncomfortable with stuff going into src/devices with these sound effect samples.  We compromised on it when it was a device in src/mame that was just a helper for making these machines a bit more “fun”, but I don’t like the idea of setting a precedent of putting these “speakers” in more devices.

Besides the philosophical aspect of it, having speakers in devices is problematic because they have spatial configuration.  The devices don’t know how they’re positioned within a system, so they inherently can’t position their speakers properly.

SomeRandomGuyIdk commented:
Addressed the code feedback. So that's what side_effects_disabled() is for :D. For the samples, could replacing the sample start() calls with reel start/stop callbacks that drivers in turn attach to something in fruit_samples_device be any better? I kinda want to keep the sounds, but no biggie if they can't make it.

DavidHaywood commented:
I think we need some way to support the sounds, and still think it would be an area where a compromise is needed.

For machines this primitive, the sounds are an important part of understanding the logic.  There are no win tunes, or anything like that, so hearing coins drop, and reels spinning is important feedback.  It can be the difference between understanding what is going on, and knowing that the machine is operating correctly, or thinking it is broken.

Even with some of the later games on hardware such as MPU4 where there are odd features like dropping a bonus coin for a certain 'square' you land on, it can be helpful to hear a coin drop sound, otherwise it isn't clear what those locations do and MAME is entirely silent about it otherwise.

DavidHaywood commented:
can we get this in before it rots please?

preferences over how samples etc. are handled can be worked out later, as they're just that, personal preferences

getting the base functionality in is important.

james-wallace-ghub commented:
Would putting this in mute resolve the impasse? If that's what it takes, I'll modify this to remove them and the original poster can review later. I agree that the core functionality here needs to be there sooner rather than later.

DavidHaywood commented:
I mean you could, but then it loses a lot of it's value.  Just feels like MAME is being needlessly stubborn over this and it's killed all momentum we were building with that scene.

SomeRandomGuyIdk commented:
Even if the delays hadn't come in I'd probably have been real busy anyway as I still am right now. JP I'd imagine as well plus he got ill a few weeks back. Right now I have MPU2 in the works, far from done with it though. I know I'm not doing some things right in this PR, just want to know what the right way is :D. The reel outputs were another complaint, would these be better off as just having the plain position get passed without scaling it, or was something else in mind there?

james-wallace-ghub commented:
Maybe you're right but it 's beyond my skillset to do this clean. @DavidHaywood, @happppp, is there any formal structure we can use here to make this work clean and in the MAME philosophy, as anything of this era is going to need some sort of mechanical sound.

happppp commented:
Me? I don't like generic samples for pinball and floppy, so my opinion on this would be the same. But I'm not involved with fruit machine emulation so I don't care that much. If let's say someone wants to add mechanical sounds for MAME's chesspieces simulation, I would block it.

What is "buzzer" for btw? Won't it differ for each hardware? type of waveform, pitch, duty cycle, or whether it's a beep or a bell sound(volume decay), etc.

SomeRandomGuyIdk commented:
Buzzer is for [this](https://www.youtube.com/watch?v=ajlJgNMMHQY&t=212s) for example, on one game it's the only source of sounds apart from the other mechanical stuff. All these samples vary by manufacturer but you can kinda do per-machine sounds by putting them under samples/[shortname] anyway, so having them more granular is possible at least.

altharic commented:
In this pull a suggestion is made to have an option in the ini file to silence the sounds if you want to maybe thats a way forward?

https://github.com/mamedev/mame/pull/11031

altharic commented:
Has this just been abandoned because the question is too hard to answer the whole point of agemame being discontinued was that mame would emulate fruit machines and now it cannot emulate fruit machines because of mame?

This was pushed in October and its now April?

james-wallace-ghub commented:
It seems to me to be the whole issue of emulation vs simulation. By nature of the beast a lot of this stuff is potentially better done in discrete audio or similar, but then we have some things that can't realistically be represented without an actual sample source. Find a way of doing that that Vasantha et al are OK with, and that will do it. It's just how, and in the meantime stuff like this is dragging on the floor.

rb6502 commented:
I'm actually unclear on how this could be done in a way that satisfies Vas.  If the objection is the sound routing, we can have a single central speaker device in the driver and route the sounds back to it.  But unless he's proposing actual physical modeling synthesis of bells and buzzers or whatnot I don't see a way around using samples.

cuavas commented:
Allowing the “fruit samples” thing in the first place was already a compromise, but it’s feeling like a “give an inch and they’ll take a mile” situation.  This is the same as why I want to take a hard line on things like #11098: you put an interface for per-game hacks in a common device, and next thing you know you have a pile of per-game hacks, no-one wants to actually characterise the device in question properly, and you’re stuck with it for decades.

Having BFM-specific accessors on a nominally generic reel device is also a bad design choice.  The thing should just allow speed profile, steps per revolution and detent positions (in terms of steps) to be configured.  The client can then define its own sensor setup in terms of the current position obtained in steps.

Sounds produced by reels spinning etc. are not comparable to things like melody chimes in EM pinballs.  Yeah, maybe at some point we’re going to want a model for simulating chimes, and samples aren’t going to cut it.  In the mean time, we *do* allow the chime solenoid outputs to be exposed as outputs, so people can use them with actual chimes (remember LEDblinky, etc. are a thing).

The other devices that already have sound samples (floppy drives) are an ongoing source of acrimony within the team.  We definitely don’t need more of that.

If you really insist on triggering samples depending on reel speed, expose a speed changed callback that the driver can subscribe to.

SomeRandomGuyIdk commented:
Copy that. I'll probably get some time to work on MAME in a few days, then I'll redo the reel device.

SomeRandomGuyIdk commented:
Cleaned up em_reel_device and moved the reel samples to fruit_samples_device, hopefully that's alright. The output steps that I mentioned remain the same.

Looking back now the questionable design choices are pretty obvious, I'll try to be less of a noob in the future :D.

altharic commented:
Is this ok to be merged now its been 3 weeks since the changes were made as requested?

angelosa commented:
You have a conflict that needs resolving first off.

altharic commented:
Any progress on the review on this?

rb6502 commented:
I *think* it now satisfies Vas's major problem of having generic devices playing samples.  The fruit-sounds device isn't ideal, but I think it's much better than having the reel device playing sounds.  I'd love to get confirmation from @cuavas on that point though.

altharic commented:
Two week later and this seems to be still stalled any update?

altharic commented:
Any progress on the review of this pull?

altharic commented:
Does anyone read these posts I have politely asked if this is being looked at and I have not even had an acknowledgement rather ignorant when all I have asked is if this is still being looked at?

happppp commented:
Your topic bumps? Nope not really, they're kind of impolite.
The ball is in cuavas's court, he can verify if his feedback was addressed. I'll give him another poke: @cuavas could you look at this again after 0.256 is out?

cuavas commented on src/mame/bfm/bfm_blackbox.cpp:
```diff
@@ -0,0 +1,1851 @@
+	m_digits.resolve();
+	m_test_led.resolve();
+
+	save_item(NAME(m_input_en));
+
+	for(int i = 0; i < 6; i++) m_input_en[i] = false;
```
Please use a range-based `for` loop or `std::fill`/`std::fill_n` in a way that doesn’t need the array size as a literal here.

johnparker007 commented:
Thanks for code reviewing this @cuavas - I fully appreciate how time-consuming (but essential for maintaining the quality) these comprehensive code reviews are :)

SomeRandomGuyIdk commented:
Alright, got it, thanks. I'll have some time to look at these next weekend.

SomeRandomGuyIdk commented:
Should be taken care of now. A few sighs were had looking at the code I wrote months ago once again. Here's hoping this saga can finally come to an end.

cuavas commented:
> Should be taken care of now. A few sighs were had looking at the code I wrote months ago once again. Here's hoping this saga can finally come to an end.

I think it should be close enough that if it isn’t ready this time around I can probably patch it up myself to a state where it can be merged.

SomeRandomGuyIdk commented:
Woohoo! Thanks a lot for the help @cuavas and for sticking through dealing with code this green, even if it ended up taking so long. I'm still getting the hang of this whole actual proper C++ thing, or at least something resembling such, but I'll try to suck less in the future 😄
